vwifi.c

```diff
@@ -404,15 +435,112 @@ static struct cfg80211_ops owl_cfg_ops = {
     .get_station = owl_get_station,
 };
 
+static int owl_ndo_open(struct net_device *dev)
+{
+    /* The first byte is '\0' to avoid being a multicast
+     * address (the first byte of multicast addrs is odd).
+     */
+    struct owl_ndev_priv_context *np = ndev_get_owl_context(dev);
+    char intf_name[ETH_ALEN] = {0};
+    snprintf(&intf_name[1], ETH_ALEN - 1, "%s", dev->name);
+    memcpy(dev->dev_addr, intf_name, ETH_ALEN);
+
+    INIT_LIST_HEAD(&np->rx_queue);
+    netif_start_queue(dev);
+    return 0;
+}
+
+static int owl_ndo_stop(struct net_device *dev)
+{
+    struct owl_ndev_priv_context *np = ndev_get_owl_context(dev);
+    struct owl_packet *pkt, *is = NULL;
+    list_for_each_entry_safe (pkt, is, &np->rx_queue, list) {
+        list_del(&pkt->list);
+        kfree(pkt);
+    }
+    netif_stop_queue(dev);
+    return 0;
+}
+
+static struct net_device_stats *owl_ndo_get_stats(struct net_device *dev)
+{
+    struct owl_ndev_priv_context *np = ndev_get_owl_context(dev);
+    return &np->stats;
+}
+
+/*
+ * Receive a packet: retrieve, encapsulate and pass over to upper levels
+ */
+static void owl_rx(struct net_device *dev)
+{
+    struct owl_ndev_priv_context *np = ndev_get_owl_context(dev);
+    struct sk_buff *skb;
+    char prefix[16];
+    struct owl_packet *pkt;
+    if (list_empty(&np->rx_queue)) {
+        printk(KERN_NOTICE "owl rx: No packet in rx_queue\n");
+        return;
+    }
+    pkt = list_first_entry(&np->rx_queue, struct owl_packet, list);
+    snprintf(prefix, 16, "Sta %s Rx ", dev->name);
+    print_hex_dump(KERN_DEBUG, prefix, DUMP_PREFIX_OFFSET, 16, 1, pkt->data,
+                   pkt->datalen, false);
+    /* Put raw packet into socket buffer */
+    skb = dev_alloc_skb(pkt->datalen + 2);
+    if (!skb) {
+        printk(KERN_NOTICE "owl rx: low on mem - packet dropped\n");
+        np->stats.rx_dropped++;
+        goto pkt_free;
+    }
+    skb_reserve(skb, 2); /* align IP on 16B boundary */
+    memcpy(skb_put(skb, pkt->datalen), pkt->data, pkt->datalen);
+    /* Write metadata, and then pass to the receive level */
+    skb->dev = dev;
+    skb->protocol = eth_type_trans(skb, dev);
+    skb->ip_summed = CHECKSUM_UNNECESSARY; /* don't check it */
+    netif_rx(skb);
+    np->stats.rx_packets++;
+    np->stats.rx_bytes += pkt->datalen;
+pkt_free:
+    list_del(&pkt->list);
+    kfree(pkt);
+}
+
 /* Network packet transmit.
  * Callback called by the kernel when packet of data should be sent.
  * In this example it does nothing.
  */
 static netdev_tx_t owl_ndo_start_xmit(struct sk_buff *skb,
                                       struct net_device *dev)
 {
+    struct owl_ndev_priv_context *np = ndev_get_owl_context(dev);
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    struct owl_ndev_priv_context *dest_np = NULL;
     /* Don't forget to cleanup skb, as its ownership moved to xmit callback. */
+    np->stats.tx_packets++;
+    np->stats.tx_bytes += skb->len;
```

Comment:
I think we should move this two lines behind the `owl_rx()`, which represents transmission finish. Or we should simulate  an  interrupt for transmission finish.

Reply:
Simulating an interrupt for transmission/receiving is a good idea.
However, pull request#15 is focused on solving redirect protocol packets to the kernel.
Let kernel handle ICMP and ARP or even the TCP packet.
Simulate the interrupt could be put to next topic after #15.

Reply:
Let's describe the further actions in GitHub issues.